TinyGo compatible packages

jphastings opened the pull request:
The `net/http/httptest` package isn't implemented in tinygo — this (minor) change swaps to creating the necessary `http.Request` struct directly (which needs an extra step or two).

**This is not enough** to get TinyGo compiled WASM _working_ for this repo, but the compilation succeeds now — one step closer!

Note: I didn't recompile all the `docs/*/.wasm` blobs with my new version of the code for this PR — there's no automation for this so I'm not sure if that's the intent. I'll be happy to make amendments to this PR as needed.

---

As an aside, any tips on how to debug what the issue is with the TinyGo compiled wasm would be appreciated! I'd love to get that — #11 — working. Try:

```sh
$ GOOS=js GOARCH=wasm tinygo build -o ./docs/hello/api.wasm  ./docs/hello/

$ serve ./docs/hello
[serve] 2024/12/14 10:08:44 http server listening at :8080
```

(This currently doesn't work — the alert never pops up, and I'm not sure how to proceed)

nlepage commented:
Hi @jphastings 
Thank you for your proposition, making the package compatible with TinyGo would be really nice.
I wanted to let you know I saw your PR, however I don’t have enough free time right now to test it and investigate why it’s not working with TinyGo.
I’ll come back to it as soon as I can.
Cheers.

jphastings commented:
No problem at all @nlepage ! Thanks for responding — no need to rush on my account 😊 

If it helps whenever you have the time — this PR should stand alone, as it is, as a PR to remove the use of the `httptest` package. It happens to help us on the path to TinyGo, but we can leave that to a different PR in the new year when  😊

nlepage commented:
Hi @jphastings 
I pushed a commit to avoid using NopCloser around the body reader.
The hello example seems to work fine with tinygo, however you have to change the URL of the wasm_exec.js used in the sw.js file of the example to use tinygo's one (see [the docs](https://tinygo.org/docs/guides/webassembly/wasm/#:~:text=Execution%20of%20the%20contents%20require%20a%20few%20JS%20helper%20functions%20which%20are%20called%20from%20WebAssembly.%20We%20have%20defined%20these%20in%20tinygo/targets/wasm_exec.js.%20It%20is%20based%20on%20%24GOROOT/misc/wasm/wasm_exec.js%20from%20the%20standard%20library%2C%20but%20is%20slightly%20different.)).
It prints a bunch of errors in the console at each request execution, which is a well known issue tinygo-org/tinygo#1140

nlepage commented:
I took some time to read how the httptest package creates the Request object.

The only thing I noticed is these protocol adjustments:

```go
req.Proto = "HTTP/1.1"
req.ProtoMinor = 1
```

This would make a lesser breaking change.

Otherwise LGTM.

On the TinyGo aspect of things, I haven't yet tested it with more elaborate examples, eg. w/ a request payload, or with server sent events, etc.

However, if it works fine, the only thing we need is a new chapter in the README explaining how to use this package w/ TinyGo.

Cheers!

jphastings commented:
Hey @nlepage  — happy new year!

I've made the HTTP Proto changes you suggested (thanks!), added a TinyGo example (with README), and updated the main readme with a few points that might get confusing with both Go/TinyGo being supported. Any thoughts?

It works well on my machine too — is there anything else you'd like me to test?

---

As an example of the size reduction; my new example hits 35% of its Go size when compiled with TinyGo 😊 

```sh
# My example compiled with Go                                                                                      
$ GOOS=js GOARCH=wasm go build -o api.wasm  .    
$ du api.wasm                                
9136    api.wasm

# My example compiled with TinyGo
$ GOOS=js GOARCH=wasm tinygo build -o api.wasm  .
$ du api.wasm                                    
3216    api.wasm
```

nlepage commented:
Hi @jphastings happy new year to you too !
Just took the time for a final review, and merged this.
Thanks a lot for your contribution, and for your patience.
I'm going to release a new version in a few minutes.
